packages/muelu/src/Utils/MueLu_AlgebraicPermutationStrategy_def.hpp

```diff
@@ -31,674 +31,690 @@
 
-    // loop over global_procOverlappingWeights and eliminate wrong entries...
-    for (size_t k = 0; k<global_procMultRequestedColIds.size(); k++) {
-      GlobalOrdinal globColId = global_procMultRequestedColIds[k];
+      // template ordinal, package (double)
+      Teuchos::reduceAll(*comm, Teuchos::REDUCE_MAX, Teuchos::as<int>(globalMultColRequests),
+                         &procMultRequestedColIds[0],
+                         &global_procMultRequestedColIds[0]);
 
-      std::vector<Scalar> MyWeightForColId(numProcs,0);
-      std::vector<Scalar> GlobalWeightForColId(numProcs,0);
+      // loop over global_procOverlappingWeights and eliminate wrong entries...
+      for (size_t k = 0; k<global_procMultRequestedColIds.size(); k++) {
+        GlobalOrdinal globColId = global_procMultRequestedColIds[k];
 
-      if(gColVec->getMap()->isNodeGlobalElement(globColId)) {
-        MyWeightForColId[myRank] = gColId2Weight[globColId];
-      } else {
-        MyWeightForColId[myRank] = SC_ZERO;
-      }
+        std::vector<MT> MyWeightForColId(numProcs, 0);
+        std::vector<MT> GlobalWeightForColId(numProcs, 0);
 
-      Teuchos::reduceAll(*comm, Teuchos::REDUCE_MAX, numProcs, &MyWeightForColId[0], &GlobalWeightForColId[0]);
-
-      if(gColVec->getMap()->isNodeGlobalElement(globColId)) {
-        // note: 2 procs could have the same weight for a column index.
-        // pick the first one.
-        typedef typename Teuchos::ScalarTraits<Scalar>::magnitudeType MT;
-        MT winnerValue = 0.0;
-        int winnerProcRank = 0;
-        for (int proc = 0; proc < numProcs; proc++) {
-          if(Teuchos::ScalarTraits<Scalar>::real (GlobalWeightForColId[proc]) > winnerValue) {
-            winnerValue = Teuchos::ScalarTraits<Scalar>::real (GlobalWeightForColId[proc]);
-            winnerProcRank = proc;
-          }
+        if(gColVec->getMap()->isNodeGlobalElement(globColId)) {
+          MyWeightForColId[myRank] = gColId2Weight[globColId];
+        } else {
+          MyWeightForColId[myRank] = MT_ZERO;
         }
 
-        // winnerProcRank is the winner for handling globColId.
-        // winnerProcRank is unique (even if two procs have the same weight for a column index)
-
-        if(myRank != winnerProcRank) {
-          // remove corresponding entry from permutedDiagCandidatesFiltered
-          typename std::vector<std::pair<GlobalOrdinal, GlobalOrdinal> >::iterator p = permutedDiagCandidatesFiltered.begin();
-          while(p != permutedDiagCandidatesFiltered.end() )
-          {
-            if((*p).second == globColId)
-              p = permutedDiagCandidatesFiltered.erase(p);
-            else
-              p++;
+        Teuchos::reduceAll(*comm, Teuchos::REDUCE_MAX, numProcs,
+                           &MyWeightForColId[0],
+                           &GlobalWeightForColId[0]);
+
+        if(gColVec->getMap()->isNodeGlobalElement(globColId)) {
+          // note: 2 procs could have the same weight for a column index.
```

Comment:
Again, this is not in scope of this PR, but Distributor and Directory might do a lot of this work for you in a more communication-efficient way.